### PR TITLE
Fuzz fix, quote invalid UTF-8 values

### PR DESCRIPTION
**Pull request by judwhite**
Summary:
- Fuzz function looked like it was testing the same input twice, I changed that and a reserialization difference was found
- The test failed for the value `"\xbd\x00"` with "reserialized data does not match"; commit message has more details
- Fixed by quoting any invalid UTF-8 string and any string containing the literal `utf8.RuneError`
- Added decode checks on the key making the parser more strict on key input. I'd like some help here, the decode benchmark difference below doesn't seem to justify the edge case it covers
- Probably not ready to merge just yet

New tests results with existing code:

```
--- FAIL: TestDecoder_errors (0.00s)
	decode_test.go:145: got: <nil>, want: logfmt syntax error at pos 5 on line 1: invalid key
	decode_test.go:145: got: <nil>, want: logfmt syntax error at pos 2 on line 1: invalid key
--- FAIL: TestEncodeKeyval (0.00s)
	encode_test.go:71: "k", "\xbd": got 'k=�', want 'k="\ufffd"'
	encode_test.go:71: "k", "�\x00": got 'k="�\u0000"', want 'k="\ufffd\u0000"'
	encode_test.go:71: "k", "�": got 'k=�', want 'k="\ufffd"'
	encode_test.go:71: "k", []byte{0xef, 0xbf, 0xbd, 0x0}: got 'k="�\u0000"', want 'k="\ufffd\u0000"'
	encode_test.go:71: "k", []byte{0xef, 0xbf, 0xbd}: got 'k=�', want 'k="\ufffd"'
--- FAIL: TestMarshalKeyvals (0.00s)
	encode_test.go:135: []interface {}{"�"}: got error: <nil>, want error: invalid key
	encode_test.go:138: []interface {}{"�"}: got '�=null', want ''
	encode_test.go:135: []interface {}{"\xbd"}: got error: <nil>, want error: invalid key
	encode_test.go:138: []interface {}{"\xbd"}: got '�=null', want ''
```

Old benchmarks:

```
BenchmarkDecodeKeyval-8              500           2454245 ns/op         203.73 MB/s
BenchmarkKRDecode-8                  500           2332233 ns/op         214.39 MB/s
BenchmarkEncodeKeyval-8          2000000               704 ns/op              64 B/op          4 allocs/op
```

New benchmarks:

```
BenchmarkDecodeKeyval-8              500           3694369 ns/op         135.34 MB/s
BenchmarkKRDecode-8                 1000           2369236 ns/op         211.04 MB/s
BenchmarkEncodeKeyval-8          2000000               716 ns/op              64 B/op          4 allocs/op
```

### CODE REVIEW

**Comment by ChrisHines** [comment]
Fantastic catch!

File: encode.go [hunk 1]
```diff
@@ -165,19 +166,27 @@ func writeKey(w io.Writer, key interface{}) error {
...
-	return r <= ' ' || r == '=' || r == '"'
+	return r <= ' ' || r == '=' || r == '"' || r == utf8.RuneError
+}
+
+func invalidKeyString(key string) bool {
+	return len(key) == 0 || strings.IndexFunc(key, invalidKeyRune) != -1 || !utf8.ValidString(key)
+}
+
+func invalidKey(key []byte) bool {
+	return len(key) == 0 || bytes.IndexFunc(key, invalidKeyRune) != -1 || !utf8.Valid(key)
```
Comment (ChrisHines) [inline encode.go]: Call to `utf8.Valid` not needed.

File: decode.go [hunk 2]
```diff
@@ -68,13 +68,19 @@ func (dec *Decoder) ScanKeyval() bool {
...
 	for p, c := range line[dec.pos:] {
 		switch {
 		case c == '=':
 			dec.pos += p
 			if dec.pos > start {
 				dec.key = line[start:dec.pos]
+				if invalidKey(dec.key) {
+					dec.syntaxError(invalidKeyError)
+					return false
+				}
```
Comment (ChrisHines) [inline decode.go]: We can avoid doing extra work for ASCII only keys.

1. Initialize `multibyte := false` outside the loop.
2. Add `case c >= utf8.RuneSelf:	multibyte = true` at the bottom of the switch.
3. Change each `if invalidKey(dec.key)` to `if multibyte && bytes.IndexRune(dec.key, utf8.RuneError) != -1`

**Comment by ChrisHines** [comment]
@judwhite Can you re-target this PR to the develop branch? This project is using Git flow now, so we shouldn't update master until we make a new release.

**Comment by judwhite** [comment]
These benchmarks were taken from a different computer.

First PR iteration benchmarks (outdated; for comparison):

```
BenchmarkDecodeKeyval-8              500           3828872 ns/op         130.59 MB/s
BenchmarkKRDecode-8                  500           2522546 ns/op         198.21 MB/s
BenchmarkEncodeKeyval-8          2000000               745 ns/op              64 B/op          4 allocs/op
```

Current `develop` benchmarks:

```
BenchmarkDecodeKeyval-8              500           2598951 ns/op         192.39 MB/s
BenchmarkKRDecode-8                  500           2491137 ns/op         200.71 MB/s
BenchmarkEncodeKeyval-8          2000000               735 ns/op              64 B/op          4 allocs/op
```

Feedback applied benchmarks:

```
BenchmarkDecodeKeyval-8              500           2802288 ns/op         178.43 MB/s
BenchmarkKRDecode-8                  500           2506353 ns/op         199.49 MB/s
BenchmarkEncodeKeyval-8          2000000               734 ns/op              64 B/op          4 allocs/op
```
